### PR TITLE
Mine and Jonna Ru's maze adventure

### PR DESCRIPTION
**Pull request by ssundblom**
Link to my deployed project:

My reflections on how this week's project turned out:

Things I'd like to have clarified or explained in more detail:

### CODE REVIEW

**Comment by ssundblom** [comment]
https://adventure-maze.netlify.app/

File: code/src/App.js [hunk 1]
```diff
@@ -1,9 +1,19 @@
...
+  const isLoading = useSelector((store) => store.ui.isLoading)
+
   return (
-    <div>
-      Find me in src/app.js!
-    </div>
+    <>
+      <Loading />
+      {!isLoading && <StartGame />}
+      {!isLoading && <GameState />}
```
Comment (johannarexin) [inline code/src/App.js:16]: Another way to do this is to write like this:

```
{!isLoading && 
<>
<StartGame />
<GameState />
</>
}
```

File: code/src/components/Action.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import React, { useState } from 'react'
+import { useSelector, useDispatch } from 'react-redux'
+import { executeAction } from '../reducers/game'
+import styled from 'styled-components'
```
Comment (johannarexin) [inline code/src/components/Action.js:4]: You should write it in this order:

import React, { useState } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import styled from 'styled-components'

import { executeAction } from '../reducers/game'

File: code/src/components/Action.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    dispatch(executeAction(username, type, direction))
+  }
+
+  return (
+    <div>
+      <h3>Go {direction} </h3>
+      <p>{description}</p>
+      <p></p>
+      <MoveButton onClick={handleOnClick}>Proceed</MoveButton>
+    </div>
```
Comment (johannarexin) [inline code/src/components/Action.js:33]: From an UX and accessibility perspective we would prefer the button to have the {direction}. So the buttons say either "Go west" or "Go north" for example. We don't think it's a good idea to have two buttons with the same text.

File: code/src/components/GameState.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import React from 'react'
+import { useSelector } from 'react-redux'
+import { Action } from "./Action"
```
Comment (johannarexin) [inline code/src/components/GameState.js:3]: It should be written like this:

import React from 'react'
import { useSelector } from 'react-redux'

import { Action } from "./Action"

File: code/src/components/GameState.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+  return (
+    <div>
+
+      <p>{username && `Player: ${username}`}</p>
+      <p>{coordinates && `Your coordinates: ${coordinates}`}</p>
+      <p>{description && description}</p>
+      <p>{actions.length > 0 && `You have ${actions.length} option(s):`}</p>
+
+      {actions && actions.map(action => {
```
Comment (johannarexin) [inline code/src/components/GameState.js:21]: Is first "actions" needed?

File: code/src/components/Loading.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import React from 'react'
+import { useSelector } from 'react-redux'
+
+
+export const Loading = () => {
+  const isLoading = useSelector((state) => state.ui.isLoading)
+  return (
+    <>{isLoading && <div>!! Loading !!</div>}</>
```
Comment (johannarexin) [inline code/src/components/Loading.js:8]: We think that "isLoading" can be removed as you have conditionals for this in App.js.

File: code/src/components/StartGame.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
+import React from 'react'
+import { useDispatch, useSelector } from 'react-redux'
+import { startNewGame } from 'reducers/game';
+import swal from "sweetalert";
+import styled from 'styled-components'
```
Comment (johannarexin) [inline code/src/components/StartGame.js:5]: Should be written like this:

import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import swal from "sweetalert";
import styled from 'styled-components'

import { startNewGame } from 'reducers/game';

File: code/src/components/StartGame.js [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+          placeholder: "What's your name, brave adventurer?",
+          type: "text",
+        },
+      },
+    }).then((value) => {
+      if (value) {
+        dispatch(startNewGame(value));
+      }
+    });
+  };
```
Comment (johannarexin) [inline code/src/components/StartGame.js:35]: Nice using swal! We have never used it before but got inspired to do so! Clean and nice code.

File: code/src/components/StartGame.js [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+  return (
+    <div>
+      <StartButton onClick={handleOnClick}>{username === '' ? 'Play' : 'Restart'}</StartButton>
```
Comment (johannarexin) [inline code/src/components/StartGame.js:39]: Very smart using the same button but with different text alternatives.

File: code/src/index.css [hunk 10]
```diff
@@ -1,13 +1,50 @@
...
+.swal-button:hover {
+  background-color: black;
+}
+
+.swal-content__input[type="text"], .swal-content__input[type="text"]:focus {
+  font-family: 'Inconsolata', monospace;
+  background-color: #000;
+  color: #4CFF42;
+  outline: #4CFF42;
+  border: 1px solid #4CFF42;
```
Comment (johannarexin) [inline code/src/index.css:44]: Why are you styling swal with regular CSS? Can't you use styled components for this?

File: code/src/reducers/game.js [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    gameState: {
+      description: '',
+      actions: []
+    }
+  },
+  reducers: {
+    setGameState: (state, action) => {
+      const newState = state
+      newState.gameState = action.payload
+      return newState
```
Comment (johannarexin) [inline code/src/reducers/game.js:17]: Why is return needed here?

File: code/src/reducers/game.js [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    },
+    setUserName: (state, action) => {
+      const newState = state
+      newState.username = action.payload
+      return newState
```
Comment (johannarexin) [inline code/src/reducers/game.js:22]: Why is return needed here?

File: code/src/reducers/reusable.js [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+  fetch('https://wk16-backend.herokuapp.com/action', {
+    method: 'POST',
+    headers: { "content-Type": "application/json" },
+    body: JSON.stringify({ username: username })
+  })
+    .then(response => response.json())
+    .then(json => {
+      setDescription(json.description)
+    })
+}
```
Comment (johannarexin) [inline code/src/reducers/reusable.js:15]: Can this be merged to executeAction?

**Review by johannarexin — CHANGES_REQUESTED** [review]
Great job, your game looks great!

We don't see that you have used media queries in your code. We think it looks great in mobile, ok on tablet but in desktop the text is jumping around since it's centered but has no width. Maybe a good idea to add it to make the site responsive (that's why we have chosen "Request changes").

Nice that you have a restart button :) And also that you have the coordinates. 
We find you code clean and easy to understand, nice job! It was fun to learn about sweetalert and that you used one button with different text alternatives :) 

/Johanna & Annika

File: code/src/components/Action.js [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+    // thunk:
+    dispatch(executeAction(username, type, direction));
+    // action:
+    dispatch(game.actions.setHistory({ type, direction }));
+  };
+
+  return (
+    <Card
+      description={description}
+      buttonText={`${type} ${direction}`}
```
Comment (puj) [inline code/src/components/Action.js:22]: Nice thinking! 💯

File: code/src/components/GameState.js [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+        )}
+        {description && <p>{description}</p>}
+        {actions.length > 0 && <h3>You have {actions.length} option(s):</h3>}
+      </TopContainer>
+
+      <BottomContainer>
+        <CardWrapper>
+          {actions &&
+            actions.map((action) => {
+              return <Action key={action.description} {...action} />;
```
Comment (puj) [inline code/src/components/GameState.js:75]: Clean

File: code/src/components/History.js [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+    <HistoryContainer>
+      {history.length > 0 && <HistoryHeading>Actions:</HistoryHeading>}
+      {history
+        .slice(-9)
+        .reverse()
+        .map((item) => (
+          <HistoryText>
+            {item.type}d {item.direction}
+          </HistoryText>
+        ))}
```
Comment (puj) [inline code/src/components/History.js:28]: `.slice(-9)` is understandable to limit the height of the column with the action history; however, from purely a code perspective the -9 might seem a bit arbitrary.   

`.slice(NUM_HISTORY_ITEMS_VISIBLE)` could be more clear.  Also, some visually indicator to the user that they are not viewing their entire history 👍

File: code/src/components/Loading.js [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+  50% {
+    content: "..";
+  }
+  75% {
+    content: "...";
+  }
+  100% {
+    content: "";
+  }
+`;
```
Comment (puj) [inline code/src/components/Loading.js:22]: Love this.

File: code/src/components/StartGame.js [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+      if (value) {
+        dispatch(startNewGame(value));
+      }
+    });
+  };
+
+  return (
+    <StyledButton play={username === ""} onClick={handleOnClick}>
+      {username === "" ? "Play" : "Restart"}
+    </StyledButton>
```
Comment (puj) [inline code/src/components/StartGame.js:32]: Prefer a constant here.  Ideally something like
```javascript
const isNewPlayer = isEmpty(username);
...

   {isNewPlayer ? "Play" : "Restart"}
```

**Review by puj — APPROVED** [review]
Clearly structured!  I like the consistency in both the code and the design.  
It's evident that you considered your solutions very carefully and iterated a few times.  

Very nice project!  Well done.

**Pull request closed without merging**